plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```
@@ -164,38 +179,39 @@ public DeleteResult delete() throws IOException {
+                () -> clientReference.get()
+                    .listObjectsV2Paginator(
+                        ListObjectsV2Request.builder()
+                            .bucket(blobStore.bucket())
+                            .prefix(keyPath)
+                            .overrideConfiguration(
+                                o -> o.addMetricPublisher(blobStore.getStatsMetricPublisher().listObjectsMetricPublisher)
+                            )
+                            .build()
+                    )
+            );
+
+            Iterator<ListObjectsV2Response> listObjectsResponseIterator = listObjectsIterable.iterator();
+            while (listObjectsResponseIterator.hasNext()) {
+                ListObjectsV2Response listObjectsResponse = SocketAccess.doPrivileged(listObjectsResponseIterator::next);
+                List<String> blobsToDelete = listObjectsResponse.contents().stream().map(s3Object -> {
                     deletedBlobs.incrementAndGet();
-                    deletedBytes.addAndGet(s3ObjectSummary.getSize());
-                    blobsToDelete.add(s3ObjectSummary.getKey());
-                });
-                if (list.isTruncated()) {
-                    doDeleteBlobs(blobsToDelete, false);
-                    prevListing = list;
-                } else {
-                    final List<String> lastBlobsToDelete = new ArrayList<>(blobsToDelete);
-                    lastBlobsToDelete.add(keyPath);
-                    doDeleteBlobs(lastBlobsToDelete, false);
-                    break;
+                    deletedBytes.addAndGet(s3Object.size());
+
+                    return s3Object.key();
+                }).collect(Collectors.toList());
+
+                if (!listObjectsResponseIterator.hasNext()) {
+                    blobsToDelete.add(keyPath);
                 }
+
+                doDeleteBlobs(blobsToDelete, false);
             }
-        } catch (final AmazonClientException e) {
+        } catch (RuntimeException e) {
```

Since the AWS hierarchy is now from `RuntimeException`, may be we could catch `SdkException` to distinguish AWS and non-AWS related issues?

---

When debugging I found that using `SdkException` swallowed all kinds of errors. I forget the repro case though. Let's leave this open for now.